src/systems/filecoin_blockchain/chainsync/_index.md

```diff
@@ -1,58 +1,388 @@
-syntactic validation is performed by the {{<sref chainsync>}},
-and once collected and validated, the blocks are forwarded
-to the {{<sref chain_manager>}} which performs semantic validation and adds the blocks to the node's current view of the blockchain state.
+Blockchain synchronization ("sync") is a key part of a blockchain system.
+It handles retrieval and propagation of blocks and transactions (messages), and
+thus in charge of distributed state replication.
+**This process is security critical -- problems here can be catastrophic to the
+operation of a blockchain.**
 
-The block syncer interfaces and functions are included here, but we expand on important details below for clarity.
+# What is ChainSync?
 
-goFile block_syncer
+`ChainSync` is the protocol Filecoin uses to synchronize its blockchain. It is
+specific to Filecoin's choices in state representation and consensus rules,
+but is general enough that it can serve other blockchains. `ChainSync` is a
+group of smaller protocols, which handle different parts of the sync process.
 
-# Block reception and Syntactic Validation
 
-- Called by: libp2p {{<sref chainsync>}}
-- Calls on: {{<sref chain_manager>}}
+# ChainSync Summary
 
-On reception of a new block over the appropriate libp2p channel (see gossib_sub), the _Block Syncer_'s _OnNewBlock_ method is invoked. Thereafter, the syncer must perform syntactic validation on the block to discard invalid blocks and forward the others for further validation by the {{<sref chain_manager>}}. At a high level, a syntactically valid block:
+At a high level, `ChainSync` does the following:
 
-- must include a well-formed miner address
-- must include at least one well-formed ticket
-- must include an election proof which is a valid signature by the miner address of the final ticket
-- must include at least one parent CID
-- must include a positive parent weight
-- must include a positive height
-- must include a well-formed state root
-- must include well-formed messages, and corresponding receipts CIDs
-- must include a valid timestamp
+- **Part 1: Verify internal state (`INIT` state below)**
+  - SHOULD verify data structures and validate local chain
+  - Resource expensive verification MAY be skipped at nodes' own risk
+- **Part 2: Bootstrap to the network (`BOOTSTRAP`)**
+  - Step 1. Bootstrap to the network, and acquire a "secure enough" set of peers (more details below)
+  - Step 2. Bootstrap to the `BlockPubsub` channels
+  - Step 3. Listen and serve on Graphsync
+- **Part 3: Synchronize trusted checkpoint state (`SYNC_CHECKPOINT`)**
+  - Step 1. Start with a `TrustedCheckpoint` (defaults to `GenesisCheckpoint`).
+  - Step 2. Get the block it points to, and that block's parents
+  - Step 3. Graphsync the `StateTree`
+- **Part 4: Catch up to the chain  (`CHAIN_CATCHUP`)**
+  - Step 1. Maintain a set of `TargetHeads`, and select the `BestTargetHead` from it
+  - Step 2. Synchronize to the latest heads observed, validating blocks towards them
```

Comment:
You want to make sure the head you selected actually links to your last checkpoint. Naively trying to sync forward towards a target that might not line up is hard (and DoSable)

Reply:
- Yes agree - and syncing back just from the head has big DOS attack surface too. [0] need to pick common intermediate targets (or use log skip lists...). haven't written the detail -- will do so.
  - anyway, In the common case, thankfully, people will download and load the chain manually along w/ the software, and wont have to worry about it. (which is why im Ok hitting testnet and even mainnet w/ this risk -- ie download headers all the way back, then validate fwd.)

[0] Downloading just headers for 1yr of a single chain is on the order of many GB. Downloading that only to find out that head is broken over and over -- reminds of old bittorrent DOS attacks of never giving the last pieces, or sending multi-GB bogus pieces that can only be detected after downloading the whole thing

Reply:
(there's a very simple + clean solution using an actor to track `TrustedCheckpoints` and a log skip list back to it, i'll write it up as a candidate for you to eval)